lfm2.5-thinking:1.2b (library tag) fails to load: feed_forward_length metadata (12288) mismatches 8192-wide FFN tensors, and final norm is mis-named output_norm

## Goal

`ollama pull lfm2.5-thinking:1.2b` (the official ollama-library tag) loads and answers coherently on
the fork — the same model that runs fine on upstream ollama today.

## Description

The official library model **`lfm2.5-thinking:1.2b`** (architecture `lfm2`, dense) will not load on
the fork. It fails the moment the weights are read, on **every** GPU including the K80 — this is a
model-load failure in the vendored llama.cpp, before any GPU kernel runs.

In plain words: the file that ollama publishes for this model has two small pieces of bookkeeping
that don't match the actual weights inside it. Upstream ollama quietly corrects both at load time, so
it runs there. Our copy of llama.cpp trusts the bookkeeping and refuses the file. We want to correct
the same two things, the way our fork does everything else to llama.cpp — as a numbered patch.

The two defects in the published GGUF (confirmed by reading the file's bytes and the loader's error):

1. **Wrong FFN width in metadata.** `lfm2.feed_forward_length = 12288`, but the real FFN tensors are
   8192-wide (`blk.0.ffn_gate.weight = [2048, 8192]`). llama.cpp allocates a `2048×12288` gate from
   the metadata, then rejects the actual `2048×8192` tensor.
2. **The final norm is mis-named.** The library GGUF ships the final embedding norm as
   `output_norm.weight` and has **no** `token_embd_norm.weight`. In llama.cpp's LFM2 graph these are
   the *same* weight (`model.tok_norm`, applied right before `model.output`). The correct Liquid
   (hf.co) GGUF names it `token_embd_norm`.

## Expected Behavior

`lfm2.5-thinking:1.2b` loads, is classified `LLM_TYPE_1_2B`, and produces coherent output — like it
does on upstream ollama, and like the correctly-converted `hf.co/LiquidAI/LFM2.5-1.2B-Thinking-GGUF`
does on the fork today.

## Steps to Reproduce

1. Any host with the current image running (no special GPU needed — fails before GPU work).
2. `docker exec ollama37 ollama pull lfm2.5-thinking:1.2b`
3. `curl -s localhost:11434/api/generate -d '{"model":"lfm2.5-thinking:1.2b","prompt":"hi","stream":false}'`

**Observed result** (ran it — this is the current #431/#437 image, i.e. *after* the token_embd_norm
fix already landed):

```
error: llama runner process has terminated: error loading model:
  check_tensor_dims: tensor 'blk.0.ffn_gate.weight' has wrong shape;
  expected  2048, 12288, got  2048,  8192,     1,     1
panic: unable to load model
```

Loader metadata dump shows `lfm2.feed_forward_length u32 = 12288` immediately before the failure.

## The path

**Model:** `lfm2.5-thinking:1.2b` — `general.architecture = lfm2` (dense).
**Hardware:** irrelevant — fails at model-load in the vendored llama.cpp, before any kernel. Observed
on the K80 (sm_37) fork image; the RTX 2060 takes the identical path.
**Config:** default; `lfm2` ∉ `OllamaEngineRequired` and has no native Go model in the fork ⇒
**llama.cpp runner**.

```
llama_model_loader: kv lfm2.feed_forward_length = 12288          (GGUF metadata, wrong)
llama-model.cpp:525    get_key_or_arr(FEED_FORWARD_LENGTH → n_ff_arr)   broadcasts 12288
llama-model.cpp:2072   switch (hparams.n_ff()) → 12288 falls to default → LLM_TYPE_UNKNOWN
llama-model.cpp:6057   create_tensor(ffn_gate, {n_embd, n_ff=12288})   expects 2048×12288
llama-model-loader.cpp:780  check_tensor_dims: actual tensor is 2048×8192 → "wrong shape" → throw
```

Second, independent defect on the same file (surfaces only once the shape check is fixed, or would
have if load reached it): the final norm ships as `output_norm.weight`, but the LFM2 loader only
looks up `token_embd_norm`. Our patch 0033 makes that norm optional-and-skipped — which for this tag
would silently drop a norm that is actually present under the other name.

## Observed vs predicted

- **Observed** (ran it): the `check_tensor_dims` "wrong shape" failure above; `feed_forward_length =
  12288` in the loader dump; direct GGUF read shows `blk.0.ffn_gate = [2048, 8192]`, `output_norm`
  present, `token_embd_norm` absent (library tag) vs the reverse on the hf.co tag.
- **Observed:** the correct `hf.co/LiquidAI/LFM2.5-1.2B-Thinking-GGUF:Q4_K_M` (ff=8192,
  `token_embd_norm` present) loads and passes on both runners today.
- **Observed:** upstream ollama corrects both via `llama/compat/llama-ollama-compat.cpp`
  `handle_lfm2()` (rename `output_norm`→`token_embd_norm`; rewrite `feed_forward_length` from the
  `blk.0.ffn_gate` width) in the model-loader constructor, before validation.
- **Predicted** (read the source, not yet run): applying the norm (rather than skipping it) yields
  coherent output — confirmation is the coherent-output judge on TC-MODELS-021.

## Blast radius

`LLM_ARCH_LFM2` **dense** only. Untested/unaffected by design: `lfm2moe` (`lfm2.5:8b` — the K80
no-harm anchor), `lfm2` 2.0 (350M/700M/2.6B), and the correct hf.co lfm2.5 GGUF — for all of these
the metadata already matches the tensors and `token_embd_norm` is present, so the fix is a guarded
no-op. No CUDA/toolchain/FA change (CUDA 11.4 / sm_37 preserved).

## Scope / Non-goals

- Fix lives in `llama/patches/0034-*.patch` (adapt to our vendored revision; **not** a cherry-pick,
  **not** a vendor bump — `docs/porting-k80.md` §1). Do **not** port upstream's whole `compat/`
  framework.
- Do not touch upstream gates (`fs/ggml/`, `ml/`) or the toolchain.
- Do not change `lfm2moe` behavior.
- Re-conversion is **not** a fix for the already-published blob (its on-disk metadata stays wrong);
  the correction must happen at load.

## Acceptance criteria

- [ ] `lfm2.5-thinking:1.2b` loads with no `check_tensor_dims` "wrong shape" and no "missing tensor",
      classified `LLM_TYPE_1_2B`, and the coherent-output judge on TC-MODELS-021 passes — on **both**
      `sm37` (K80) and `sm75` (RTX 2060).
- [ ] **K80 no-harm gate:** `lfm2moe` (`lfm2.5:8b`) throughput within **±3%** of the `d83061de`
      baseline (~33 tok/s flat), per `docs/reports/k80-vram-by-family-2026-07-10-d83061de.md`
      methodology (FA off, f16 KV). This fix should be a no-op there — make it a criterion.
- [ ] The change is `llama/patches/0034-*.patch` and re-applies cleanly via `Makefile.sync`.
- [ ] The correct hf.co lfm2.5 GGUF and `lfm2` 2.0 still load unchanged (guarded no-op).

## Unverified

- Semantic equivalence of the library's `output_norm` and llama.cpp's `token_embd_norm` (treated as
  the same final norm; the coherence judge is the proof).
- Whether the library GGUF's `tokenizer.ggml.pre` is correct — upstream's shim also sets it to
  `lfm2`; this fix does not. If wrong, tokenization could degrade coherence independently; check the
  load log on first run and add a small additive override if needed.

## Links

- Related: #386 (token_embd_norm required) and its task #431 / PR #437 — those addressed only the
  *first-seen* symptom (norm absent). This issue delivers #386's still-unmet acceptance,
  "`lfm2.5-thinking:1.2b` loads and produces coherent output," by fixing the two real blockers.
- Precedent patch: `llama/patches/0033-lfm2-token_embd_norm-optional.patch`.

--- a/cicd/tests/testcases/models/TC-MODELS-021.yml
+++ b/cicd/tests/testcases/models/TC-MODELS-021.yml
@@ -4,13 +4,15 @@ suite: models
 priority: 2
 timeout: 600000
 dependencies: []
-issue: ""
+issue: "438"
 intent:
   user_story: |
     Validate lfm2.5-thinking:1.2b (Liquid LFM2, thinking, ~0.7 GB) runs on K80 compute 3.7
-    (single GPU). Acceptance for STORY-019. KNOWN RED: on the current image the GGUF fails to
-    load with "missing tensor 'token_embd_norm.weight'" — this gates that vendored-llama.cpp
-    LFM2 fix. "think": false pinned so the answer isn't buried in a <think> span.
+    (single GPU). The ollama-library GGUF is mis-converted — feed_forward_length metadata
+    (12288) disagrees with its 8192-wide FFN tensors, and its final norm ships as output_norm
+    (no token_embd_norm); patch 0034 loads it by deriving n_ff from the ffn_gate tensor and
+    resolving tok_norm from output_norm. "think": false pinned so the answer isn't buried in a
+    <think> span.
   notes: |
     Prerequisite: lfm2.5-thinking:1.2b pre-pulled on the runner.
     Acceptable warning (NOT an error): "flash attention enabled but not supported by gpu".
@@ -27,6 +29,7 @@ steps:
       - "CUBLAS_STATUS"
       - "CUDA error"
       - "missing tensor"
+      - "wrong shape"
 
   - name: Check GPU memory
     command: docker exec ollama37 nvidia-smi --query-compute-apps=pid,used_memory --format=csv
@@ -62,5 +65,6 @@ steps:
 
 criteria: |
   Validate lfm2.5-thinking:1.2b (Liquid LFM2 thinking, ~0.7 GB) runs on K80 (compute 3.7),
-  single die. Prerequisite: pre-pulled. KNOWN RED until the vendored-llama.cpp LFM2 loader
-  handles token_embd_norm.weight for this variant (observed load failure).
+  single die. Prerequisite: pre-pulled. Loads via patch 0034 (#438) — feed_forward_length
+  corrected from the ffn_gate tensor width, final norm resolved from output_norm; no
+  "wrong shape" or "missing tensor" load failure.

--- a/llama/llama.cpp/src/llama-arch.cpp
+++ b/llama/llama.cpp/src/llama-arch.cpp
@@ -2189,6 +2189,11 @@ static const std::map<llm_arch, std::map<llm_tensor, const char *>> LLM_TENSOR_N
             { LLM_TENSOR_SHORTCONV_OUTPROJ, "blk.%d.shortconv.out_proj" },
             { LLM_TENSOR_TOKEN_EMBD,        "token_embd" },
             { LLM_TENSOR_TOKEN_EMBD_NORM,   "token_embd_norm" },
+            // The ollama-library lfm2.5 GGUF names the final embedding norm "output_norm"
+            // (llama.cpp's LFM2 graph uses it as model.tok_norm). Register it so the loader
+            // can resolve tok_norm from output_norm. LFM2 dense only — intentionally absent
+            // from the LFM2MOE map below.
+            { LLM_TENSOR_OUTPUT_NORM,       "output_norm" },
             { LLM_TENSOR_OUTPUT,            "output" },
         }
     },

--- a/llama/llama.cpp/src/llama-model.cpp
+++ b/llama/llama.cpp/src/llama-model.cpp
@@ -2069,6 +2069,26 @@ void llama_model::load_hparams(llama_model_loader & ml) {
                     hparams.recurrent_layer_arr[il] = hparams.n_head_kv(il) == 0;
                 }
                 hparams.n_layer_dense_lead = hparams.n_layer;
+
+                // The ollama-library lfm2.5-thinking:1.2b GGUF ships a wrong
+                // lfm2.feed_forward_length (12288) that disagrees with its actual dense FFN
+                // tensors (blk.0.ffn_gate.weight = [n_embd, 8192]). Trust the tensor: derive
+                // the real n_ff from tensor metadata and overwrite n_ff_arr so BOTH the type
+                // switch below and the load_tensors expected dim (hparams.n_ff()) match the
+                // weights. No-op when metadata already agrees (correct hf.co lfm2.5 and lfm2
+                // 2.0 GGUFs), so it cannot disturb them. Mirrors upstream ollama's handle_lfm2.
+                if (const ggml_tensor * t_gate = ml.get_tensor_meta("blk.0.ffn_gate.weight")) {
+                    const int64_t real_n_ff = t_gate->ne[1];
+                    if (real_n_ff > 0 && (uint32_t) real_n_ff != hparams.n_ff()) {
+                        LLAMA_LOG_WARN("%s: lfm2 feed_forward_length metadata (%u) disagrees with "
+                                       "blk.0.ffn_gate.weight width (%lld); using tensor width\n",
+                                       __func__, hparams.n_ff(), (long long) real_n_ff);
+                        for (uint32_t il = 0; il < hparams.n_layer; ++il) {
+                            hparams.n_ff_arr[il] = (uint32_t) real_n_ff;
+                        }
+                    }
+                }
+
                 switch (hparams.n_ff()) {
                     case  4608: type = LLM_TYPE_350M; break;
                     case  6912: type = LLM_TYPE_700M; break;
@@ -6033,6 +6053,14 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
                     // config ships no token_embd_norm — so make it optional and skip the norm in
                     // llm_build_lfm2 when absent. (lfm2 2.0 ships it; the guard keeps those intact.)
                     tok_norm = create_tensor(tn(LLM_TENSOR_TOKEN_EMBD_NORM, "weight"), {n_embd}, TENSOR_NOT_REQUIRED);
+                    // The ollama-library lfm2.5 GGUF ships the final embedding norm as
+                    // output_norm.weight (no token_embd_norm). In llm_build_lfm2 this is the SAME
+                    // weight (model.tok_norm, applied right before model.output), so fall back to
+                    // output_norm rather than dropping a real norm. Dense-LFM2 only; output_norm is
+                    // unregistered for LFM2MOE, so this is a no-op there even without the guard.
+                    if (tok_norm == NULL && arch == LLM_ARCH_LFM2) {
+                        tok_norm = create_tensor(tn(LLM_TENSOR_OUTPUT_NORM, "weight"), {n_embd}, TENSOR_NOT_REQUIRED);
+                    }
                     output   = create_tensor(tn(LLM_TENSOR_OUTPUT,          "weight"), {n_embd, n_vocab}, TENSOR_NOT_REQUIRED);
 
                     if (output == NULL) {

--- a/llama/patches/0034-lfm2-library-gguf-load.patch
+++ b/llama/patches/0034-lfm2-library-gguf-load.patch
@@ -1,0 +1,93 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Shang Chieh Tseng <shangchieh.tseng@tsengsyu.com>
+Date: Mon, 13 Jul 2026 11:30:44 +0800
+Subject: [PATCH] fix(lfm2): load the ollama-library lfm2.5-thinking:1.2b GGUF
+
+The ollama-library lfm2.5-thinking:1.2b GGUF (arch lfm2, dense) has two defects
+vs a correct conversion, and fails to load in vendored llama.cpp:
+
+1. lfm2.feed_forward_length metadata is 12288 but the actual dense FFN tensors
+   are 8192-wide (blk.0.ffn_gate.weight = [2048, 8192]) -> check_tensor_dims
+   rejects it. Derive the real n_ff from the ffn_gate tensor and overwrite
+   n_ff_arr (fixes both the type switch and the load-time expected dim).
+2. the final embedding norm ships as output_norm.weight (no token_embd_norm).
+   In llm_build_lfm2 this is the same weight (model.tok_norm). Register
+   output_norm in the LFM2 (dense) arch map and fall back to it when
+   token_embd_norm is absent, so the norm is APPLIED, not skipped by patch 0033.
+
+Corrects patch 0033's premise: lfm2.5 does have the embedding norm -- shipped as
+output_norm (ollama-library GGUF) / token_embd_norm (hf.co GGUF), not absent.
+
+Scoped to LLM_ARCH_LFM2 dense and guarded on a genuine mismatch, so lfm2moe
+(lfm2.5:8b), lfm2 2.0, and the correct hf.co lfm2.5 GGUF are unchanged. Mirrors
+upstream ollama's llama/compat handle_lfm2 behavior. Arch/loading only -- no
+CUDA/toolchain change (K80 unaffected).
+
+Co-Authored-By: Claude Opus 4.8 (1M context) <noreply@anthropic.com>
+---
+ src/llama-arch.cpp  |  5 +++++
+ src/llama-model.cpp | 28 ++++++++++++++++++++++++++++
+ 2 files changed, 33 insertions(+)
+
+diff --git a/src/llama-arch.cpp b/src/llama-arch.cpp
+index 9f6b6ad20..738550321 100644
+--- a/src/llama-arch.cpp
++++ b/src/llama-arch.cpp
+@@ -2129,6 +2129,11 @@ static const std::map<llm_arch, std::map<llm_tensor, const char *>> LLM_TENSOR_N
+             { LLM_TENSOR_SHORTCONV_OUTPROJ, "blk.%d.shortconv.out_proj" },
+             { LLM_TENSOR_TOKEN_EMBD,        "token_embd" },
+             { LLM_TENSOR_TOKEN_EMBD_NORM,   "token_embd_norm" },
++            // The ollama-library lfm2.5 GGUF names the final embedding norm "output_norm"
++            // (llama.cpp's LFM2 graph uses it as model.tok_norm). Register it so the loader
++            // can resolve tok_norm from output_norm. LFM2 dense only — intentionally absent
++            // from the LFM2MOE map below.
++            { LLM_TENSOR_OUTPUT_NORM,       "output_norm" },
+             { LLM_TENSOR_OUTPUT,            "output" },
+         }
+     },
+diff --git a/src/llama-model.cpp b/src/llama-model.cpp
+index c7ec1a0b7..0c84dc3b8 100644
+--- a/src/llama-model.cpp
++++ b/src/llama-model.cpp
+@@ -2021,6 +2021,26 @@ void llama_model::load_hparams(llama_model_loader & ml) {
+                     hparams.recurrent_layer_arr[il] = hparams.n_head_kv(il) == 0;
+                 }
+                 hparams.n_layer_dense_lead = hparams.n_layer;
++
++                // The ollama-library lfm2.5-thinking:1.2b GGUF ships a wrong
++                // lfm2.feed_forward_length (12288) that disagrees with its actual dense FFN
++                // tensors (blk.0.ffn_gate.weight = [n_embd, 8192]). Trust the tensor: derive
++                // the real n_ff from tensor metadata and overwrite n_ff_arr so BOTH the type
++                // switch below and the load_tensors expected dim (hparams.n_ff()) match the
++                // weights. No-op when metadata already agrees (correct hf.co lfm2.5 and lfm2
++                // 2.0 GGUFs), so it cannot disturb them. Mirrors upstream ollama's handle_lfm2.
++                if (const ggml_tensor * t_gate = ml.get_tensor_meta("blk.0.ffn_gate.weight")) {
++                    const int64_t real_n_ff = t_gate->ne[1];
++                    if (real_n_ff > 0 && (uint32_t) real_n_ff != hparams.n_ff()) {
++                        LLAMA_LOG_WARN("%s: lfm2 feed_forward_length metadata (%u) disagrees with "
++                                       "blk.0.ffn_gate.weight width (%lld); using tensor width\n",
++                                       __func__, hparams.n_ff(), (long long) real_n_ff);
++                        for (uint32_t il = 0; il < hparams.n_layer; ++il) {
++                            hparams.n_ff_arr[il] = (uint32_t) real_n_ff;
++                        }
++                    }
++                }
++
+                 switch (hparams.n_ff()) {
+                     case  4608: type = LLM_TYPE_350M; break;
+                     case  6912: type = LLM_TYPE_700M; break;
+@@ -5894,6 +5914,14 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+                     // config ships no token_embd_norm — so make it optional and skip the norm in
+                     // llm_build_lfm2 when absent. (lfm2 2.0 ships it; the guard keeps those intact.)
+                     tok_norm = create_tensor(tn(LLM_TENSOR_TOKEN_EMBD_NORM, "weight"), {n_embd}, TENSOR_NOT_REQUIRED);
++                    // The ollama-library lfm2.5 GGUF ships the final embedding norm as
++                    // output_norm.weight (no token_embd_norm). In llm_build_lfm2 this is the SAME
++                    // weight (model.tok_norm, applied right before model.output), so fall back to
++                    // output_norm rather than dropping a real norm. Dense-LFM2 only; output_norm is
++                    // unregistered for LFM2MOE, so this is a no-op there even without the guard.
++                    if (tok_norm == NULL && arch == LLM_ARCH_LFM2) {
++                        tok_norm = create_tensor(tn(LLM_TENSOR_OUTPUT_NORM, "weight"), {n_embd}, TENSOR_NOT_REQUIRED);
++                    }
+                     output   = create_tensor(tn(LLM_TENSOR_OUTPUT,          "weight"), {n_embd, n_vocab}, TENSOR_NOT_REQUIRED);
+ 
+                     if (output == NULL) {

